src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java

```diff
@@ -187,57 +202,159 @@ private void dataCmd() {
+    }
+    if (isMultiPart(thePart)) {
+      nextHandler.handle(null);
+    } else {
+      // send single part with the endHandler
+      sendRegularPart(thePart, nextHandler);
+    }
+  }
+
+  private boolean isMultiPart(EncodedPart part) {
+    return part.parts() != null && part.parts().size() > 0;
+  }
+
+  private void sendMailHeaders(EncodedPart part) {
+    for (Map.Entry<String, String> entry: part.headers()) {
+      connection.writeLine(entry.toString(), wrottern.getAndAdd(entry.toString().length()) < 1000);
+    }
+    // send empty line between headers and body
+    connection.writeLine("", wrottern.get() < 1000);
+  }
+
+  private void sendRegularPart(EncodedPart part, Handler<Void> endHandler) {
+    sendMailHeaders(part);
+    if (part.body() != null) {
+      // send body string
+      for (String bodyLine: part.body().split("\n")) {
+        String line = bodyLine.startsWith(".") ? "." + bodyLine : bodyLine;
+        connection.writeLine(line, wrottern.getAndAdd(line.length()) < 1000);
+      }
+      endHandler.handle(null);
+    } else if (part.bodyStream() != null) {
+      BodyReadStream bodyReadStream = new BodyReadStream(part.bodyStream());
+      bodyReadStream.pipe().endOnComplete(false).to(connection.getSocket(), r -> {
+        if (r.succeeded()) {
+          endHandler.handle(null);
+        } else {
+          log.warn("Failed to transfer stream", r.cause());
+          handleError(r.cause());
+        }
+      });
+    } else {
+      throw new IllegalStateException("No mail body and stream found");
+    }
+  }
+
+  private Handler<Void> handlerInContext(Handler<Void> handler) {
+    return vv -> connection.getContext().runOnContext(v -> handler.handle(null));
+  }
+
+  // what we need: strings line by line with CRLF as line terminator
+  private class BodyReadStream implements ReadStream<Buffer> {
+
+    private final ReadStream<Buffer> stream;
+    private final int size = 57;
+    private Buffer streamBuffer;
+    private Handler<Buffer> handler;
+
+    private BodyReadStream(ReadStream<Buffer> stream) {
+      Objects.requireNonNull(stream, "ReadStream cannot be null");
+      this.stream = stream;
+      this.streamBuffer = Buffer.buffer();
+    }
+
+    @Override
+    public BodyReadStream exceptionHandler(Handler<Throwable> handler) {
+      if (handler != null) {
+        stream.exceptionHandler(handler);
+      }
+      return this;
+    }
+
+    @Override
+    public BodyReadStream handler(@Nullable Handler<Buffer> handler) {
+      if (handler == null) {
+        return this;
+      }
+      this.handler = handler;
+      stream.handler(b -> connection.getContext().runOnContext(v -> {
+        Buffer buffer = streamBuffer.appendBuffer(b);
+        int start = 0;
+        while(start + size < buffer.length()) {
+          final String theLine = Utils.base64(buffer.getBytes(start, start + size));
+          handler.handle(Buffer.buffer(theLine + "\r\n"));
```

Comment:
we need to call the `handler` only a single time because this stream implementation delegates the `fetch` operation and does not calculate the number of fetches. For instance if this stream is fetched `1` but the provided `Buffer` is actually split in `4` then the handler will be called `4` times although it only asked for `1` buffer.

Reply:
Ah, that is the thing I missed, thank you. I will fix it.

Reply:
yes this is a bit tricky